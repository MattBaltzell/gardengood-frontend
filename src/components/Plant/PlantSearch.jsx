import React, { useState, useEffect } from "react";
import "./Plant.css";
import "../Form/Form.css";
import PlantList from "./PlantList";

const PlantSearch = ({ plants, search }) => {
  const INITIAL_STATE = { searchTerm: "" };
  const [formData, setFormData] = useState(INITIAL_STATE);

  const handleChange = async (e) => {
    const { name, value } = e.target;
    setFormData((data) => ({ ...data, [name]: value }));
    search(value.trim() || undefined);
  };

  useEffect(() => {
    return () => {
      setFormData(INITIAL_STATE);
    };
  }, []);

  const handleSubmit = (e) => {
    e.preventDefault();
    search(formData.searchTerm.trim() || undefined);
  };

  return (
    <main className="Plant-search">
      <h1>Find a Plant</h1>
      <form className="Form Plant-search__form" onSubmit={handleSubmit}>
        <input
          className="Form__input--text"
          name="searchTerm"
          onChange={handleChange}
          type="text"
          value={formData.searchTerm}
          placeholder="Search plant name"
        />
      </form>
      <PlantList plants={plants} />
    </main>
  );
};

export default PlantSearch;
