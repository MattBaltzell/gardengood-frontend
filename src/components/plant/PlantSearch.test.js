import React from "react";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import PlantSearch from "./PlantSearch";
import UserContext from "../auth/UserContext";

const user = {
  username: "testuser",
  first_name: "firstname",
  last_name: "lastname",
  email: "test@test.com",
};

const handleIsLoadingMock = jest.fn();

describe("PlantSearch component tests", function () {
  it("renders without crashing", function () {
    render(
      <MemoryRouter>
        <PlantSearch />
      </MemoryRouter>
    );
  });

  it("matches snapshot", function () {
    const { asFragment } = render(
      <MemoryRouter>
        <UserContext.Provider value={user}>
          <PlantSearch />
        </UserContext.Provider>
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot when logged out", function () {
    const { asFragment } = render(
      <MemoryRouter>
        <UserContext.Provider value={null}>
          <PlantSearch
            isLoading={false}
            handleIsLoading={handleIsLoadingMock}
          />
        </UserContext.Provider>
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders loading message when loading", function () {
    render(
      <MemoryRouter>
        <UserContext.Provider value={user}>
          <PlantSearch isLoading={true} handleIsLoading={handleIsLoadingMock} />
        </UserContext.Provider>
      </MemoryRouter>
    );
    expect(screen.getByText("Loading...")).toBeInTheDocument();
  });

  it("renders plants list", async function () {
    render(
      <MemoryRouter>
        <UserContext.Provider value={user}>
          <PlantSearch
            isLoading={false}
            handleIsLoading={handleIsLoadingMock}
          />
        </UserContext.Provider>
      </MemoryRouter>
    );
    expect(await screen.findByText("Test Plant 1")).toBeInTheDocument();
    expect(await screen.findByText("Test Plant 2")).toBeInTheDocument();
  });
});
