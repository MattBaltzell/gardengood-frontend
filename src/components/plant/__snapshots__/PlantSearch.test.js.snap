// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlantSearch component tests matches snapshot 1`] = `
<DocumentFragment>
  <main
    class="Plant-search"
  >
    <h1>
      Plant Directory
    </h1>
    <form
      autocomplete="off"
      class="Plant-search__form"
    >
      <input
        autocomplete="false"
        name="hidden"
        style="display: none;"
        type="text"
      />
      <input
        class="Form__input--text"
        name="searchTerm"
        placeholder="Start typing plant name"
        type="text"
        value=""
      />
    </form>
    <div
      aria-busy="true"
      aria-label="ball-triangle-loading"
      class="[object Object]"
      data-testid="ball-triangle-loading"
      role="status"
      style="display: flex;"
    >
      <svg
        data-testid="ball-triangle-svg"
        height="100"
        stroke="var(--color-primary-400)"
        viewBox="0 0 57 57"
        width="100"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fill-rule="evenodd"
        >
          <g
            stroke-width="2"
            transform="translate(1 1)"
          >
            <circle
              cx="5"
              cy="50"
              r="5"
            >
              <animate
                attributeName="cy"
                begin="0s"
                calcMode="linear"
                dur="2.2s"
                repeatCount="indefinite"
                values="50;5;50;50"
              />
              <animate
                attributeName="cx"
                begin="0s"
                calcMode="linear"
                dur="2.2s"
                repeatCount="indefinite"
                values="5;27;49;5"
              />
            </circle>
            <circle
              cx="27"
              cy="5"
              r="5"
            >
              <animate
                attributeName="cy"
                begin="0s"
                calcMode="linear"
                dur="2.2s"
                from="5"
                repeatCount="indefinite"
                to="5"
                values="5;50;50;5"
              />
              <animate
                attributeName="cx"
                begin="0s"
                calcMode="linear"
                dur="2.2s"
                from="27"
                repeatCount="indefinite"
                to="27"
                values="27;49;5;27"
              />
            </circle>
            <circle
              cx="49"
              cy="50"
              r="5"
            >
              <animate
                attributeName="cy"
                begin="0s"
                calcMode="linear"
                dur="2.2s"
                repeatCount="indefinite"
                values="50;50;5;50"
              />
              <animate
                attributeName="cx"
                begin="0s"
                calcMode="linear"
                dur="2.2s"
                from="49"
                repeatCount="indefinite"
                to="49"
                values="49;5;27;49"
              />
            </circle>
          </g>
        </g>
      </svg>
    </div>
    <p
      class="pulsing"
    >
      Loading...
    </p>
  </main>
</DocumentFragment>
`;

exports[`PlantSearch component tests matches snapshot when logged out 1`] = `
<DocumentFragment>
  <main
    class="Plant-search"
  >
    <h1>
      Plant Directory
    </h1>
    <form
      autocomplete="off"
      class="Plant-search__form"
    >
      <input
        autocomplete="false"
        name="hidden"
        style="display: none;"
        type="text"
      />
      <input
        class="Form__input--text"
        name="searchTerm"
        placeholder="Start typing plant name"
        type="text"
        value=""
      />
    </form>
    <div
      aria-busy="true"
      aria-label="ball-triangle-loading"
      class="[object Object]"
      data-testid="ball-triangle-loading"
      role="status"
      style="display: flex;"
    >
      <svg
        data-testid="ball-triangle-svg"
        height="100"
        stroke="var(--color-primary-400)"
        viewBox="0 0 57 57"
        width="100"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fill-rule="evenodd"
        >
          <g
            stroke-width="2"
            transform="translate(1 1)"
          >
            <circle
              cx="5"
              cy="50"
              r="5"
            >
              <animate
                attributeName="cy"
                begin="0s"
                calcMode="linear"
                dur="2.2s"
                repeatCount="indefinite"
                values="50;5;50;50"
              />
              <animate
                attributeName="cx"
                begin="0s"
                calcMode="linear"
                dur="2.2s"
                repeatCount="indefinite"
                values="5;27;49;5"
              />
            </circle>
            <circle
              cx="27"
              cy="5"
              r="5"
            >
              <animate
                attributeName="cy"
                begin="0s"
                calcMode="linear"
                dur="2.2s"
                from="5"
                repeatCount="indefinite"
                to="5"
                values="5;50;50;5"
              />
              <animate
                attributeName="cx"
                begin="0s"
                calcMode="linear"
                dur="2.2s"
                from="27"
                repeatCount="indefinite"
                to="27"
                values="27;49;5;27"
              />
            </circle>
            <circle
              cx="49"
              cy="50"
              r="5"
            >
              <animate
                attributeName="cy"
                begin="0s"
                calcMode="linear"
                dur="2.2s"
                repeatCount="indefinite"
                values="50;50;5;50"
              />
              <animate
                attributeName="cx"
                begin="0s"
                calcMode="linear"
                dur="2.2s"
                from="49"
                repeatCount="indefinite"
                to="49"
                values="49;5;27;49"
              />
            </circle>
          </g>
        </g>
      </svg>
    </div>
    <p
      class="pulsing"
    >
      Loading...
    </p>
  </main>
</DocumentFragment>
`;
